src/client/address_book_storage.h

```diff
@@ -0,0 +1,176 @@
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.               //
+ *                                                                                            //
+ * Parts of the project are originally copyright (c) 2013-2015 The PurpleI2P Project          //
+ */
+
+#ifndef SRC_CLIENT_ADDRESS_BOOK_STORAGE_H_
+#define SRC_CLIENT_ADDRESS_BOOK_STORAGE_H_
+
+#include <boost/filesystem.hpp>
+
+#include <cstdint>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "identity.h"
+#include "core/util/filesystem.h"
+#include "router_context.h"
+#include "util/base64.h"
+#include "util/log.h"
+
+namespace i2p {
+namespace client {
+
+/// @class AddressBookDefaults
+/// @brief Default string constants used throughout address book
+struct AddressBookDefaults {
+  /// @brief Pathname of address book within KOVRI_DATA_DIR
+  /// @return Pathname for address book storage
+  std::string GetDefaultPathname() const {
+    return "addressbook";
+  };
+
+  /// @brief Gets default publishers filename
+  /// @return Default publishers filename
+  /// @notes A publishers file holds a list of publisher addresses
+  ///   of whom publish 'subscriptions' that contain a list of hosts to .b32.i2p
+  std::string GetDefaultPublishersFilename() const {
+    return "publishers.txt";
+  }
+
+  /// @brief Gets default publisher URI
+  /// @return Default publishers URI
+  /// @notes A default publisher is used if no publishers file is available
+  std::string GetDefaultPublisherURI() const {
+    // TODO(unassigned): replace with Monero's b32 publisher service
+    return "https://downloads.getmonero.org/kovri/hosts.txt";
+    // Below is only used for testing in-net download (this is *not* our default subscription)
+    //return "http://udhdrtrcetjm5sxzskjyr5ztpeszydbh4dpl3pl4utgqqw2v4jna.b32.i2p/hosts.txt";
+  }
+
+  /// @brief Gets default subscription filename
+  /// @return Default subscription filename
+  /// @notes Filename used by publishers when they publish a 'subscription'
+  std::string GetDefaultSubscriptionFilename() const {
+    return "hosts.txt";
+  }
+
+  /// @brief Gets addresses file (file list of saved addresses)
+  /// @return Default addresses filename
+  /// @notes Currently only used to verify that addresses have indeed been saved
+  std::string GetDefaultAddressesFilename() const {
+    return "addresses.csv";
+  }
+};
+
+/// @class AddressBookStorage
+/// @brief Interface for address book storage
+class AddressBookStorage {
+ public:
+  virtual ~AddressBookStorage() {}
+
+  virtual bool GetAddress(
+      const i2p::data::IdentHash& ident,
+      i2p::data::IdentityEx& address) const = 0;
+
+  virtual void AddAddress(
+      const i2p::data::IdentityEx& address) = 0;
+
+  /**
+  // TODO(unassigned): currently unused
+  virtual void RemoveAddress(
+      const i2p::data::IdentHash& ident) = 0;
+  **/
+
+  virtual std::size_t Load(
```

Comment:
This was pre-existing. AFAICT, as an interface, it may have been going in the direction of being reused elsewhere (its old code, who knows). I don't see why we would do that now, I'll look closer.